File: TOMIMSV4/DAL/Mantenimientos/Propietario/Destinatario/clsLnPropietario_destinatario.vb
Imports System.Configuration
Imports System.Data.SqlClient

Public Class clsLnPropietario_destinatario

    Public Shared Sub Cargar(ByRef oBePropietario_destinatario As clsBePropietario_destinatario, ByRef dr As DataRow)
        Try

            With oBePropietario_destinatario
                .IdDestinatarioPropietario = IIf(IsDBNull(dr.Item("IdDestinatarioPropietario")), 0, dr.Item("IdDestinatarioPropietario"))
                .IdPropietario = IIf(IsDBNull(dr.Item("IdPropietario")), 0, dr.Item("IdPropietario"))
                .Nombre = IIf(IsDBNull(dr.Item("nombre")), "", dr.Item("nombre"))
                .Apellido = IIf(IsDBNull(dr.Item("apellido")), "", dr.Item("apellido"))
                .Correo_electronico = IIf(IsDBNull(dr.Item("correo_electronico")), "", dr.Item("correo_electronico"))
                .Telefono = IIf(IsDBNull(dr.Item("telefono")), "", dr.Item("telefono"))
                .Telefono1 = IIf(IsDBNull(dr.Item("telefono1")), "", dr.Item("telefono1"))
                .Cargo = IIf(IsDBNull(dr.Item("cargo")), "", dr.Item("cargo"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With

        Catch ex As Exception
            Throw ex
        End Try

    End Sub

    Public Function Insertar(ByRef oBePropietario_destinatario As clsBePropietario_destinatario,
                             Optional ByVal pConection As SqlConnection = Nothing,
                             Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Ins.Init("propietario_destinatario")
            Ins.Add("iddestinatariopropietario", "@iddestinatariopropietario", DataType.Parametro)
            Ins.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            Ins.Add("apellido", "@apellido", DataType.Parametro)
            Ins.Add("correo_electronico", "@correo_electronico", DataType.Parametro)
            Ins.Add("telefono", "@telefono", DataType.Parametro)
            Ins.Add("telefono1", "@telefono1", DataType.Parametro)
            Ins.Add("cargo", "@cargo", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)

            Dim sp As String = Ins.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDDESTINATARIOPROPIETARIO", oBePropietario_destinatario.IdDestinatarioPropietario))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBePropietario_destinatario.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBePropietario_destinatario.Nombre))
            cmd.Parameters.Add(New SqlParameter("@APELLIDO", oBePropietario_destinatario.Apellido))
            cmd.Parameters.Add(New SqlParameter("@CORREO_ELECTRONICO", oBePropietario_destinatario.Correo_electronico))
            cmd.Parameters.Add(New SqlParameter("@TELEFONO", oBePropietario_destinatario.Telefono))
            cmd.Parameters.Add(New SqlParameter("@TELEFONO1", oBePropietario_destinatario.Telefono1))
            cmd.Parameters.Add(New SqlParameter("@CARGO", oBePropietario_destinatario.Cargo))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBePropietario_destinatario.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
        End Try

    End Function

    Public Function Actualizar(ByRef oBePropietario_destinatario As clsBePropietario_destinatario, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand

        Try

            Upd.Init("propietario_destinatario")
            Upd.Add("iddestinatariopropietario", "@iddestinatariopropietario", DataType.Parametro)
            Upd.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("apellido", "@apellido", DataType.Parametro)
            Upd.Add("correo_electronico", "@correo_electronico", DataType.Parametro)
            Upd.Add("telefono", "@telefono", DataType.Parametro)
            Upd.Add("telefono1", "@telefono1", DataType.Parametro)
            Upd.Add("cargo", "@cargo", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdDestinatarioPropietario = @IdDestinatarioPropietario")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDDESTINATARIOPROPIETARIO", oBePropietario_destinatario.IdDestinatarioPropietario))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBePropietario_destinatario.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBePropietario_destinatario.Nombre))
            cmd.Parameters.Add(New SqlParameter("@APELLIDO", oBePropietario_destinatario.Apellido))
            cmd.Parameters.Add(New SqlParameter("@CORREO_ELECTRONICO", oBePropietario_destinatario.Correo_electronico))
            cmd.Parameters.Add(New SqlParameter("@TELEFONO", oBePropietario_destinatario.Telefono))
            cmd.Parameters.Add(New SqlParameter("@TELEFONO1", oBePropietario_destinatario.Telefono1))
            cmd.Parameters.Add(New SqlParameter("@CARGO", oBePropietario_destinatario.Cargo))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBePropietario_destinatario.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
        End Try

    End Function

    Public Shared Function Eliminar(ByRef oBePropietario_destinatario As clsBePropietario_destinatario, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text

            Dim sp As String = " Delete from Propietario_destinatario" &
             "  Where(IdDestinatarioPropietario = @IdDestinatarioPropietario)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDDESTINATARIOPROPIETARIO", oBePropietario_destinatario.IdDestinatarioPropietario))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Propietario_destinatario"

            Dim lConnection As New SqlConnection(ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim dad As New SqlDataAdapter(cmd)

            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBePropietario_destinatario As clsBePropietario_destinatario) As Boolean

        Try

            Dim sp As String = "SELECT * FROM Propietario_destinatario" &
            " Where(IdDestinatarioPropietario = @IdDestinatarioPropietario)"

            Dim lConnection As New SqlConnection(ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDDESTINATARIOPROPIETARIO", oBePropietario_destinatario.IdDestinatarioPropietario))
            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBePropietario_destinatario, dt.Rows(0))
            Else
                Throw New Exception("No se pudo obtener el registro")
            End If

            Return True

        Catch ex As Exception
            Throw ex
        End Try

    End Function

End Class

File: WMSWebAPI/Dal/Propietario/clsLnPropietario_destinatario.cs
using System.Data;
using System.Diagnostics;
using System.Reflection;
using Microsoft.Data.SqlClient;
using Microsoft.VisualBasic.CompilerServices;
using WMSWebAPI.Entity.Propietario;
public class clsLnPropietario_destinatario
{

    private static clsInsert Ins = new clsInsert();
    private static clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBePropietario_destinatario oBePropietario_destinatario, DataRow dr)
    {
        int GetInt(string col) { return dr[col] is DBNull ? 0 : Convert.ToInt32(dr[col]); }
        bool GetBool(string col) { return dr[col] is DBNull ? false : Convert.ToBoolean(dr[col]); }
        string GetString(string col) { return dr[col] is DBNull ? "" : (Convert.ToString(dr[col]) ?? ""); }
        
        try
        {
            oBePropietario_destinatario.IdDestinatarioPropietario = GetInt("IdDestinatarioPropietario");
            oBePropietario_destinatario.IdPropietario = GetInt("IdPropietario");
            oBePropietario_destinatario.Nombre = GetString("nombre");
            oBePropietario_destinatario.Apellido = GetString("apellido");
            oBePropietario_destinatario.Correo_electronico = GetString("correo_electronico");
            oBePropietario_destinatario.Telefono = GetString("telefono");
            oBePropietario_destinatario.Telefono1 = GetString("telefono1");
            oBePropietario_destinatario.Cargo = GetString("cargo");
            oBePropietario_destinatario.Activo = GetBool("activo");
        }
        catch (Exception ex)
        {
            var st = new System.Diagnostics.StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = sf?.GetMethod();
            string vMsgError = string.Format("{{0}} {{1}}", currentMethodName, ex.Message);            
            throw new Exception(vMsgError);
        }
    }

    public static int Insertar(IConfiguration config, clsBePropietario_destinatario oBePropietario_destinatario, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        int rowsAffected = 0;        
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("propietario_destinatario");
            Ins.Add("iddestinatariopropietario", "@iddestinatariopropietario", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("apellido", "@apellido", "F");
            Ins.Add("correo_electronico", "@correo_electronico", "F");
            Ins.Add("telefono", "@telefono", "F");
            Ins.Add("telefono1", "@telefono1", "F");
            Ins.Add("cargo", "@cargo", "F");
            Ins.Add("activo", "@activo", "F");

            string sp = Ins.SQL();

            var cmd = new SqlCommand(sp, lConnection) { CommandType = (CommandType)Conversions.ToInteger(CommandType.Text) };

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            Bind(cmd,oBePropietario_destinatario);

            rowsAffected = cmd.ExecuteNonQuery();

            cmd.Dispose();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection is not null) lConnection.Dispose();
            if (lTransaction is not null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public static int Insertar(IConfiguration config, clsBePropietario_destinatario oBePropietario_destinatario)
    {

        int rowsAffected = 0;
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("propietario_destinatario");
            Ins.Add("iddestinatariopropietario", "@iddestinatariopropietario", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("apellido", "@apellido", "F");
            Ins.Add("correo_electronico", "@correo_electronico", "F");
            Ins.Add("telefono", "@telefono", "F");
            Ins.Add("telefono1", "@telefono1", "F");
            Ins.Add("cargo", "@cargo", "F");
            Ins.Add("activo", "@activo", "F");

            string sp = Ins.SQL();

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            cmd = new SqlCommand(sp, lConnection, lTransaction);

            Bind(cmd, oBePropietario_destinatario);

            rowsAffected = cmd.ExecuteNonQuery();

            if (lTransaction != null)
                lTransaction.Commit();

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public static int Actualizar(IConfiguration config, clsBePropietario_destinatario oBePropietario_destinatario, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        int rowsAffected = 0;
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            Upd.Init("propietario_destinatario");
            Upd.Add("iddestinatariopropietario", "@iddestinatariopropietario", "F");
            Upd.Add("idpropietario", "@idpropietario", "F");
            Upd.Add("nombre", "@nombre", "F");
            Upd.Add("apellido", "@apellido", "F");
            Upd.Add("correo_electronico", "@correo_electronico", "F");
            Upd.Add("telefono", "@telefono", "F");
            Upd.Add("telefono1", "@telefono1", "F");
            Upd.Add("cargo", "@cargo", "F");
            Upd.Add("activo", "@activo", "F");
            Upd.Where("IdDestinatarioPropietario = @IdDestinatarioPropietario");

            string sp = Upd.SQL();

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            Bind(cmd, oBePropietario_destinatario);

            rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();


        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public int Eliminar(IConfiguration config, clsBePropietario_destinatario oBePropietario_destinatario, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = (" Delete from Propietario_destinatario" +
             "  Where(IdDestinatarioPropietario = @IdDestinatarioPropietario)");

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            cmd.Parameters.Add(new SqlParameter("@IdDestinatarioPropietario", oBePropietario_destinatario.IdDestinatarioPropietario));

            int rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return rowsAffected;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }    

    public static bool GetSingle(IConfiguration config, ref clsBePropietario_destinatario pBePropietario_destinatario)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            const string sp = "Select * FROM Propietario_destinatario" +
            " Where(IdDestinatarioPropietario = @IdDestinatarioPropietario)";

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);

            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);

            dad.SelectCommand.Parameters.Add(new SqlParameter("@IdDestinatarioPropietario", pBePropietario_destinatario.IdDestinatarioPropietario));            

            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            if (dt.Rows.Count == 1)
            {
                DataRow r;
                r = dt.Rows[0];
                Cargar(ref pBePropietario_destinatario, r);
                return true;
            }

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return false;

    }

    public static List<clsBePropietario_destinatario> GetAll(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;
        List<clsBePropietario_destinatario> lreturnList = new List<clsBePropietario_destinatario>();

        try
        {
            const string sp = "Select * FROM Propietario_destinatario";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {

                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlDataAdapter lDTA = new SqlDataAdapter(sp, lConnection))
                    {
                        lDTA.SelectCommand.CommandType = CommandType.Text;
                        lDTA.SelectCommand.Transaction = lTransaction;
                        DataTable lDataTable = new DataTable();
                        lDTA.Fill(lDataTable);

                        clsBePropietario_destinatario vBePropietario_destinatario = new clsBePropietario_destinatario();

                        foreach (DataRow dr in lDataTable.Rows)
                        {
                            vBePropietario_destinatario = new clsBePropietario_destinatario();
                            Cargar(ref vBePropietario_destinatario, dr);
                            lreturnList.Add(vBePropietario_destinatario);
                        }

                        lTransaction.Commit();
                    }

                    lConnection.Close();

                }

            }

            return lreturnList;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
    }

    public static int MaxID(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;

        try
        {

            int lMax = 0;

            const string sp = "Select ISNULL(Max(IdDestinatarioPropietario),0) FROM Propietario_destinatario";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {
                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlCommand lCommand = new SqlCommand(sp, lConnection) { CommandType = CommandType.Text })
                    {
                        Object lreturnValue = lCommand.ExecuteScalar();
                        if (lreturnValue != DBNull.Value && lreturnValue != null)
                        {
                            lMax = int.Parse((String)lreturnValue);
                        }
                    }
                    lTransaction.Commit();
                }

                lConnection.Close();
            }

            return lMax;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
    }
    public static int MaxID(IConfiguration config, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;
        int lMax = 0;
        try
        {


            const string sp = "Select ISNULL(Max(IdDestinatarioPropietario),0) FROM Propietario_destinatario";

            bool Es_Transaccion_Remota = pConection is not null && pTransaction is not null;
            var cmd = new SqlCommand(sp, lConnection) { CommandType = (CommandType)Conversions.ToInteger(CommandType.Text) };
            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            Object lreturnValue = cmd.ExecuteScalar();

            if (lreturnValue != DBNull.Value && lreturnValue != null)
            {
                lMax = int.Parse((String)lreturnValue);
            }

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return lMax;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);
            
            throw new Exception(vMsgError);
        }
    }
    private static void Bind(SqlCommand cmd, clsBePropietario_destinatario oBe)
    {
        cmd.Parameters.Add(new SqlParameter("@IdDestinatarioPropietario", oBe.IdDestinatarioPropietario));
        cmd.Parameters.Add(new SqlParameter("@IdPropietario", oBe.IdPropietario));
        cmd.Parameters.Add(new SqlParameter("@nombre", oBe.Nombre ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@apellido", oBe.Apellido ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@correo_electronico", oBe.Correo_electronico ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@telefono", oBe.Telefono ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@telefono1", oBe.Telefono1 ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@cargo", oBe.Cargo ?? (object)DBNull.Value));
        cmd.Parameters.Add(new SqlParameter("@activo", oBe.Activo));
    }

}
